### PR TITLE
Please review

### PR DESCRIPTION
**Pull request by apesternikov**
Hi Carlos,
This is gossie modified to use
https://github.com/apesternikov/thrift4go
with updated compiler producing much cleaner typed semantics for lists, maps and sets, as well as faster and cleaner library.

Also, added gomock - generated cassandra mock, should be useful for writing unit tests.

### CODE REVIEW

**Comment by carloscm** [comment]
Hello Aleksey,

First of all thank you for your work on Gossie, and also in Thrift4Go.

I've reviewed your changes for Gossie and I have some comments.
- You've changed some import paths for both Thrift and the Cassandra generated code to point to your Github repos. While this is OK while developing I prefer not to merge such changes into my master branch. When you make a pull request for Gossie please make sure you fix the Cassandra Github path to point to my repo, and the Thrift4Go to point to pomack's. Of course if in the future either pomack or myself stop maintaining our repos then it would be OK.
- I've reviewed the actual code changes and they appear safe, just adaptations to the new container styles in your changes to Thrift. Good work.
- As for the changes in Thrift this is where I believe most of the potential problems could come from. For this reason I will wait for pomack to merge your pull request into Thrift4Go before merging your changes into Gossie. Then we can test it (if possible starting from Cassandra 1.0 or even 0.8), and if there are no regressions, merge it.
- For Gossie I've preferred not to use mocks of any kind, given the focus of the library, which is in isolating the many nuances of the Cassandra Thrift interface from the developer. Mocking this interface would defeat this purpose. I am not generally opposed to the concept of mocking, but for Gossie, I believe testing directly over an actual live instance of Cassandra gives much more value than using mocks.

In conclusion I like your changes, specially your work in Thrift4Go, but I prefer to wait until your changes are merged in pomack's repo before changing anything in Gossie. We can try to coordinate both merges so there is a minimun window of breakage when they happen (but keep in mind the issue with the import paths I mentioned earlier).

Thank you!

**Comment by apesternikov** [comment]
Absolutely agree with all points with only one note re mocking.
mock is not a replacement for testing against real cassandra instance. It is to check that your layer is preparing specific requests for specific operations. For example, the mapper is using binary marshalling for struct fields, and tests against cassandra instance wouldn't break if due to bug this marshalling changes to string or flip endianness. Or to make sure that you are using proper ConsistencyLevel constants - you have no chance to check it otherwise. Third thing from the top of my head - testing situations difficult to reproduce on a real instance, like retries on timeouts and reconnects. I certainly forgot something, I'm not a testing guru.
